File: android/lib/src/main/java/org/elastos/sdk/elephantwallet/contact/internal/ContactBridge.java
package org.elastos.sdk.elephantwallet.contact.internal;

import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import org.elastos.sdk.elephantwallet.contact.Contact;
import org.elastos.tools.crosspl.CrossBase;
import org.elastos.tools.crosspl.annotation.CrossClass;
import org.elastos.tools.crosspl.annotation.CrossInterface;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

@CrossClass
public class ContactBridge extends CrossBase {
    public static final String TAG = "elastos";

    protected ContactBridge() {
        super(ContactBridge.class.getName(), 0);
    }

    public void finalize() {
        if(mListener != null) {
            mListener.unbind();
        }

        super.finalize();
    }


    public synchronized void setListener(Contact.Listener listener) {
        if(mListener != null) {
            mListener.unbind();
        }
        mListener = listener;

        mListener.bind();
        setListener(mListener);
    }

    public int setUserInfo(UserInfo.Item item, String value) {
        int ret = setUserInfo(item.id(), value);
        return ret;
    }

    public int setIdentifyCode(IdentifyCode.Type type, String value) {
        int ret = setIdentifyCode(type.id(), value);
        return ret;
    }

    public Contact.UserInfo getUserInfo() {
        assert(mListener != null);

        StringBuffer sbInfo = new StringBuffer();
        int ret = getHumanInfo("-user-info-", sbInfo);
        if(ret < 0) {
            Log.w(TAG, "Failed to get user info. ret=" + ret);
            return null;
        }

        Contact.UserInfo userInfo = new Contact.UserInfo();
        ret  = userInfo.fromJson(sbInfo.toString());
        if(ret < 0) {
            Log.w(TAG, "Failed to deserialize user info. ret=" + ret);
            return null;
        }

        return userInfo;
    }

    public List<Contact.FriendInfo> listFriendInfo() {
        assert(mListener != null);

        StringBuffer json = new StringBuffer();
        int ret = getFriendList(json);
        if(ret < 0) {
            Log.w(TAG, "Failed to list friend info. ret=" + ret);
            return null;
        }

        TypeToken<List<Contact.FriendInfo.FriendJson>> friendInfoListType = new TypeToken<List<Contact.FriendInfo.FriendJson>>(){};
        List<Contact.FriendInfo.FriendJson> listJson = new Gson().fromJson(json.toString(), friendInfoListType.getType());

        List<Contact.FriendInfo> list = new ArrayList<>();
        for(Contact.FriendInfo.FriendJson it: listJson) {
            Contact.FriendInfo friendInfo = new Contact.FriendInfo();
            ret  = friendInfo.fromJson(it);
            if(ret < 0) {
                Log.w(TAG, "Failed to deserialize friend info. ret=" + ret);
                return null;
            }

            list.add(friendInfo);
        }

        return list;
    }

    public List<String> listFriendCode() {
        List<String> friendCodeList = new ArrayList<String>();

        List<Contact.FriendInfo> friendList = listFriendInfo();
        if(friendList == null) {
            return friendCodeList;
        }

        for(Contact.FriendInfo it: friendList) {
            friendCodeList.add(it.humanCode);
        }

        return friendCodeList;
    }

    public ContactStatus getStatus(String humanCode) {
        int ret = getHumanStatus(humanCode);
        if(ret < 0) {
            return null;
        }

        return ContactStatus.valueOf(ret);
    }

    public int sendMessage(String friendCode, ContactChannel channelType, Contact.Message message) {
        if(message == null) {
            return -1;
        }

        int ret = message.syncMessageToNative();
        if(ret < 0) {
            return ret;
        }

        ret = sendMessage(friendCode, channelType.id(), message);
        if(ret < 0) {
            return ret;
        }

        return ret;
    }

    public int pullFile(String friendCode, ContactChannel channelType,
                        Contact.Message.FileData fileInfo) {
        if(fileInfo == null) {
            return -1;
        }

        int ret = pullData(friendCode, channelType.id(), fileInfo.devId, fileInfo.toString());

        return ret;
    }

    public static Contact.Message MakeTextMessage(String text, String cryptoAlgorithm) {
        Contact.Message msg = new Contact.Message(text, cryptoAlgorithm);
        return msg;
    }

    public static Contact.Message MakeFileMessage(File file, String cryptoAlgorithm) {
        Contact.Message msg = new Contact.Message(file, cryptoAlgorithm);
        return msg;
    }

//    private static Contact.Message MakeMessage(ContactMessage.Type type, byte[] data, String cryptoAlgorithm) {
//        Contact.Message msg = new Contact.Message(type, data, cryptoAlgorithm);
//        return msg;
//    }

    @CrossInterface
    public native int start();

    @CrossInterface
    public native int stop();

    @CrossInterface
    public native int addFriend(String friendCode, String summary);

    @CrossInterface
    public native int removeFriend(String friendCode);

    @CrossInterface
    public native int acceptFriend(String friendCode);

    @CrossInterface
    public native int syncInfoDownloadFromDidChain();

    @CrossInterface
    public native int syncInfoUploadToDidChain();

    @CrossInterface
    public native int setWalletAddress(String name, String value);

    @CrossInterface
    private native void setListener(CrossBase listener);

    @CrossInterface
    private native int setUserInfo(int item, String value);

    @CrossInterface
    private native int setIdentifyCode(int item, String value);

    @CrossInterface
    private native int getHumanInfo(String humanCode, StringBuffer info);

    @CrossInterface
    private native int getFriendList(StringBuffer info);

    @CrossInterface
    private native int getHumanStatus(String humanCode);

    @CrossInterface
    private native int sendMessage(String friendCode, int channelType, CrossBase message);

    @CrossInterface
    private native int pullData(String friendCode, int channelType, String devId, String fileInfo);

    private ContactListener mListener;

    static {
        System.loadLibrary("Elastos.SDK.Contact.Jni");
    }
}

File: ios/sdk/swift/internal/ContactBridge.swift
import CrossPL

/* @CrossClass */
open class ContactBridge: CrossBase {
  public static let TAG = "elastos"
  
  init() {
    super.init(className: String(describing: ContactBridge.self))
  }

  deinit {
    if(mListener != nil) {
      mListener!.unbind()
    }
  }
  
  public func setListener(listener: Contact.Listener?) {
    if(mListener != nil) {
      mListener!.unbind();
    }
    mListener = listener;

    mListener!.bind();
    setListener(listener: mListener);
  }

  public func setUserInfo(item: UserInfo.Item, value: String) -> Int {
    let ret = setUserInfo(item: item.rawValue, value: value);
    return ret
  }

  public func setIdentifyCode(type: IdentifyCode.Kind, value: String) -> Int {
    let ret = setIdentifyCode(type: type.rawValue, value: value);
    return ret;
  }

  public func getUserInfo() -> Contact.UserInfo? {
    if(mListener == nil) {
      return nil
    }
    
    var sbInfo = String()
    var ret = getHumanInfo(humanCode: "-user-info-", info: &sbInfo)
    if(ret < 0) {
      Log.w(tag: ContactBridge.TAG, msg: "Failed to get user info. ret=\(ret)")
      return nil
    }

    let userInfo = Contact.UserInfo();
    ret = userInfo.fromJson(info: sbInfo);
    if(ret < 0) {
      Log.w(tag: ContactBridge.TAG, msg: "Failed to deserialize user info. ret=\(ret)")
      return nil
    }

    return userInfo;
  }

  public func listFriendInfo() -> [Contact.FriendInfo]? {
    assert(mListener != nil)

    var json = String()
    let ret = getFriendList(info: &json);
    if(ret < 0) {
      Log.w(tag: ContactBridge.TAG, msg: "Failed to list friend info. ret=\(ret)")
      return nil
    }

//    TypeToken<List<Contact.FriendInfo.FriendJson>> friendInfoListType = new TypeToken<List<Contact.FriendInfo.FriendJson>>(){};
//    let listJson = new Gson().fromJson(json.toString(), friendInfoListType.getType());
    let listJson = try! JSONDecoder().decode([Contact.FriendInfo.FriendJson].self, from: json.data(using: .utf8)!)

    var list = [Contact.FriendInfo]()
    for it in listJson {
      let friendInfo = Contact.FriendInfo()
      let ret = friendInfo.fromJson(info: it)
      if(ret < 0) {
        Log.w(tag: ContactBridge.TAG, msg: "Failed to deserialize friend info. ret=\(ret)")
        return nil
      }

      list.append(friendInfo)
    }

    return list
  }

  public func listFriendCode() -> [String] {
    var friendCodeList = [String]()

    let friendList = listFriendInfo()
    if(friendList == nil) {
      return friendCodeList
    }

    for it in friendList! {
      friendCodeList.append(it.humanCode!)
    }

    return friendCodeList
  }

  public func getStatus(humanCode: String) -> ContactStatus? {
    let ret = getHumanStatus(humanCode: humanCode)
    if(ret < 0) {
      return nil
    }

    return ContactStatus(rawValue: ret)
  }

  public func makeMessage(type: ContactMessage.Kind, data: Data, cryptoAlgorithm: String?) -> Contact.Message {
    let msg = Contact.Message(type: type, data: data, cryptoAlgorithm: cryptoAlgorithm)
    return msg
  }

  public func makeTextMessage(data: String, cryptoAlgorithm: String?) -> Contact.Message {
    let msg = makeMessage(type: ContactMessage.Kind.MsgText,
                          data: data.data(using: .utf8)!,
                          cryptoAlgorithm: cryptoAlgorithm)
    return msg
  }


  public func sendMessage(friendCode: String, channelType: ContactChannel, message: Contact.Message) -> Int {
    var ret = message.syncMessageToNative()
    if(ret < 0) {
      return ret;
    }

    ret = sendMessage(friendCode: friendCode, channelType: channelType.rawValue, message: message)

    return ret;
  }
  
  /* @CrossNativeInterface */
  public func start() -> Int {
    let ret = crosspl_Proxy_ContactBridge_start(nativeHandle)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func stop() -> Int {
    let ret = crosspl_Proxy_ContactBridge_stop(nativeHandle)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func addFriend(friendCode: String, summary: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_addFriend(nativeHandle, friendCode, summary)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func removeFriend(friendCode: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_removeFriend(nativeHandle, friendCode)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func acceptFriend(friendCode: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_acceptFriend(nativeHandle, friendCode)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func syncInfoDownloadFromDidChain() -> Int {
    let ret = crosspl_Proxy_ContactBridge_syncInfoDownloadFromDidChain(nativeHandle)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func syncInfoUploadToDidChain() -> Int {
    let ret = crosspl_Proxy_ContactBridge_syncInfoUploadToDidChain(nativeHandle)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func setWalletAddress(name: String, value: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_setWalletAddress(nativeHandle, name, value)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  private func setListener(listener: CrossBase?) {
    crosspl_Proxy_ContactBridge_setListener(nativeHandle, listener)
  }
  
  /* @CrossNativeInterface */
  private func setUserInfo(item: Int, value: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_setUserInfo(nativeHandle, Int32(item), value)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  private func setIdentifyCode(type: Int, value: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_setIdentifyCode(nativeHandle, Int32(type), value)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  private func getHumanInfo(humanCode: String, info: inout String) -> Int {
    var nsinfo = info as NSString?
    
    let ret = crosspl_Proxy_ContactBridge_getHumanInfo(nativeHandle, humanCode, &nsinfo)
    if ret < 0 {
      return Int(ret)
    }
    info = nsinfo! as String
    
    return 0
  }
  
  /* @CrossNativeInterface */
  private func getFriendList(info: inout String) -> Int {
    var nsinfo = info as NSString?
    let ret = crosspl_Proxy_ContactBridge_getFriendList(nativeHandle, &nsinfo)
    if ret < 0 {
      return Int(ret)
    }
    info = nsinfo! as String
    
    return 0
  }
  
  /* @CrossNativeInterface */
  private func getHumanStatus(humanCode: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_getHumanStatus(nativeHandle, humanCode)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  private func sendMessage(friendCode: String, channelType: Int, message: CrossBase) -> Int {
    let ret = crosspl_Proxy_ContactBridge_sendMessage(nativeHandle, friendCode, Int32(channelType), message)
    return Int(ret)
  }
  
  private var mListener: CrossBase?
}
